File: judge/java/tests/bst_to_dlist.java
package tests;
import java.util.*;
import java.lang.*;
import java.io.*;
import judge.*;
import datastruct.*;

public class bst_to_dlist {
    public static int num_test = 100;
    public static TreeNode[] in_0;
    public static TreeNode[] in_org_0;
    public static int[][] out;

    public static boolean check_dlist(int[] arr, TreeNode list) {
        TreeNode cur = list;
        TreeNode pre = null;
        int i = 0;
        while(cur != null) {
            if(i >= arr.length || arr[i] != cur.val) return false;
            ++i;
            pre = cur;
            cur = cur.right;
        }
        if(i != arr.length) return false;
        cur = pre;
        --i;
        while(cur != null) {
            if(i < 0 || arr[i] != cur.val) return false;
            --i;
            TreeNode last = cur;
            cur = cur.left;
            last.left = null;
        }
        if(i != -1) return false;
        return true;
    }

    public static String dlist_to_string(TreeNode list) {
        String ret = "[";
        boolean first = true;
        while(list != null) {
            if(!first) ret += ", ";
            ret += Integer.toString(list.val);
            list = list.right;
            first = false;
        }
        ret += "]";
        return ret;
    }

    public static void load_test() {
        File fil = new File("judge/tests/bst-to-dlist.txt");
        FileReader inputFil = null;
        try {
            inputFil = new FileReader(fil);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        BufferedReader in = new BufferedReader(inputFil);
        try {
            in_0 = common.read_tree_array(in);
            in_org_0 = test_common.copy(in_0);
            out = common.read_int_matrix(in);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void judge() {
        load_test();
        Solution s = new Solution();

        long startTime = System.currentTimeMillis();

        for(int i = 0; i < num_test; ++i) {
            TreeNode answer = s.bst_to_list(in_0[i]);
            if(!check_dlist(out[i], answer)) {
                System.out.printf("%d / %d;", i+1, num_test);
                String outs = common.to_string(bst_to_dlist.in_org_0[i]);
                System.out.print(outs + ";");
                System.out.print(dlist_to_string(answer) + ";");
                System.out.println(common.to_string(out[i]));
                return;
            }
        }

        long estimatedTime = System.currentTimeMillis() - startTime;
        System.out.print("Accepted;");
        System.out.println(estimatedTime);
    }
}

File: judge/lua/tests/bst-to-dlist.lua
require("../solution")

local num_test = 100
local in_0 = {}
local in_org_0 = {}
local out = {}

function check_dlist(arr, list)
	local cur, pre, i = list, nil, 1

	while cur ~= nil do
		if i > #arr or arr[i] ~= cur.val then
		    return false
		end
		i = i + 1
		pre = cur
		cur = cur.right
    end

	if i ~= #arr + 1 then
	    return false
	end

	cur = pre
	i = i - 1

	while cur ~= nil do
		if (i <= 0 or arr[i] ~= cur.val) then
		    return false
		end
		i = i - 1
		last = cur
		cur = cur.left
		last.left = nil
	end

	if i ~= 0 then
	    return false
	end

	return true
end

function dlist_to_string(list)
	local ret = "["
	local first = true
	while list ~= nil do
		if not first then
		    ret = ret .. ", "
		end
		ret = ret .. list.val
		list = list.right
		first = false
  end
	return ret .. "]"
end

function load_test()
    local f = io.open("./judge/tests/bst-to-dlist.txt", "r")
    in_0 = read_tree_array(f)
    in_org_0 = copy(in_0)
    out = read_num_matrix(f)
    f:close()
end

function judge()
    load_test()
    capture_stdout()

    local start = os.clock()
    for i = 1, num_test do
        print("Testing case #" .. i)
        local answer = bst_to_list(in_0[i]) 
        if not check_dlist(out[i], answer) then
            io.write(string.format("%d / %d;", i, num_test))
            io.write(to_string(in_org_0[i]))
            io.write(";")
            io.write(dlist_to_string(answer))
            io.write(";")
            io.write(to_string(out[i]))
            io.write("\n")
            return
        end
    end

    release_stdout()
    local elapsed = math.floor((os.clock() - start) * 1000)
	print("Accepted;" .. elapsed)
end
